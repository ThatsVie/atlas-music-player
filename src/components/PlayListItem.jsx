import React from 'react';
import PropTypes from 'prop-types';

const PlayListItem = ({ title, artist, duration, isCurrent }) => {
  return (
    <div
      className={`flex justify-between items-center p-3 rounded-md ${
        isCurrent ? 'bg-teal' : 'hover:bg-brightYellow'
      }`}
    >
      <div>
        <p
          className={`text-sm font-bold ${
            isCurrent ? 'text-white' : 'text-darkMagenta'
          }`}
        >
          {title}
        </p>
        <p
          className={`text-xs ${
            isCurrent ? 'text-citrusOrange' : 'text-gray-900'
          }`}
        >
          {artist}
        </p>
      </div>

      <p className={`text-xs ${isCurrent ? 'text-citrusOrange' : 'text-gray-900'}`}>
        {duration}
      </p>
    </div>
  );
};

PlayListItem.propTypes = {
  title: PropTypes.string.isRequired,
  artist: PropTypes.string.isRequired,
  duration: PropTypes.string.isRequired,
  isCurrent: PropTypes.bool,
};

PlayListItem.defaultProps = {
  isCurrent: false,
};

export default PlayListItem;
