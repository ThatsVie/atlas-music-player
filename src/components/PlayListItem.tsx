import React from 'react';

interface PlayListItemProps {
  id: string;
  title: string;
  artist: string;
  duration: string;
  isCurrent: boolean;
  onClick: () => void;
  onEnter: () => void;
}

const PlayListItem: React.FC<PlayListItemProps> = ({
  id,
  title,
  artist,
  duration,
  isCurrent,
  onClick,
  onEnter,
}) => {
  return (
    <li
      id={id}
      className={`flex justify-between items-center p-2 rounded-md cursor-pointer ${
        isCurrent
          ? 'bg-teal dark:bg-lightTeal text-black dark:text-black'
          : 'hover:bg-brightYellow dark:hover:bg-darkMagenta transition-colors'
      }`}
      onClick={onClick}
      onKeyDown={(e) => {
        if (e.key === 'Enter') {
          onEnter();
        }
      }}
      tabIndex={0}
      role='option'
      aria-selected={isCurrent} 
    >
      <div>
        <p
          className={`text-sm font-bold ${
            isCurrent
              ? 'text-black dark:text-black'
              : 'text-darkMagenta dark:text-lightTeal'
          }`}
        >
          {title}
        </p>
        <p
          className={`text-xs ${
            isCurrent
              ? 'text-black dark:text-black'
              : 'text-gray-700 dark:text-lightTeal'
          }`}
        >
          {artist}
        </p>
      </div>
      <p
        className={`text-xs ${
          isCurrent
            ? 'text-black dark:text-black'
            : 'text-gray-700 dark:text-lightTeal'
        }`}
      >
        {duration}
      </p>
    </li>
  );
};

export default PlayListItem;
