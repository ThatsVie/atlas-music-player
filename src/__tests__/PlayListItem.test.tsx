import { render, fireEvent, screen } from '@testing-library/react';
import { describe, test, expect, vi, afterEach } from 'vitest';
import PlayListItem from '@/components/PlayListItem';

describe('PlayListItem Component', () => {
  const mockOnClick = vi.fn();
  const mockOnEnter = vi.fn();

  afterEach(() => {
    vi.clearAllMocks();
  });

  test('renders correctly with all props', () => {
    // Arrange & Act
    const { container } = render(
      <PlayListItem
        id="unique-1"
        title="Song Title"
        artist="Artist Name"
        duration="3:45"
        isCurrent={false}
        onClick={mockOnClick}
        onEnter={mockOnEnter}
      />
    );

    // Assert
    expect(container).toMatchSnapshot();
  });

  test('applies correct styles when isCurrent is true', () => {
    // Arrange
    render(
      <PlayListItem
        id="unique-2"
        title="Current Song"
        artist="Current Artist"
        duration="4:20"
        isCurrent={true}
        onClick={mockOnClick}
        onEnter={mockOnEnter}
      />
    );

    // Act
    const listItem = screen.getByRole('option', { name: /Current Song/i });

    // Assert
    expect(listItem.classList.contains('bg-teal')).toBe(true);
    expect(listItem.classList.contains('dark:bg-lightTeal')).toBe(true);
    expect(listItem.classList.contains('text-black')).toBe(true);
  });

  test('calls onClick when clicked', () => {
    // Arrange
    render(
      <PlayListItem
        id="unique-3"
        title="Click Test"
        artist="Click Artist"
        duration="2:30"
        isCurrent={false}
        onClick={mockOnClick}
        onEnter={mockOnEnter}
      />
    );

    // Act
    const listItem = screen.getByRole('option', { name: /Click Test/i });
    fireEvent.click(listItem);

    // Assert
    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  test('calls onEnter when Enter key is pressed', () => {
    // Arrange
    render(
      <PlayListItem
        id="unique-4"
        title="Song 1"
        artist="Artist 1"
        duration="3:00"
        isCurrent={false}
        onClick={mockOnClick}
        onEnter={mockOnEnter}
      />
    );

    // Act
    const listItem = screen.getByRole('option', { name: /Song 1/i });
    fireEvent.keyDown(listItem, { key: 'Enter', code: 'Enter' });

    // Assert
    expect(mockOnEnter).toHaveBeenCalledTimes(1);
  });

  test('does not call onEnter for other keys', () => {
    // Arrange
    render(
      <PlayListItem
        id="unique-5"
        title="Song 2"
        artist="Artist 2"
        duration="4:00"
        isCurrent={false}
        onClick={mockOnClick}
        onEnter={mockOnEnter}
      />
    );

    // Act
    const listItem = screen.getByRole('option', { name: /Song 2/i });
    fireEvent.keyDown(listItem, { key: 'Space', code: 'Space' });

    // Assert
    expect(mockOnEnter).not.toHaveBeenCalled();
  });
});
