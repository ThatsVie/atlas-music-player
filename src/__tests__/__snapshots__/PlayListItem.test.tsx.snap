// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayListItem Component > renders correctly with all props 1`] = `
<div>
  <li
    aria-selected="false"
    class="flex justify-between items-center p-2 rounded-md cursor-pointer hover:bg-brightYellow dark:hover:bg-darkMagenta transition-colors"
    id="unique-1"
    role="option"
    tabindex="0"
  >
    <div>
      <p
        class="text-sm font-bold text-darkMagenta dark:text-lightTeal"
      >
        Song Title
      </p>
      <p
        class="text-xs text-gray-700 dark:text-lightTeal"
      >
        Artist Name
      </p>
    </div>
    <p
      class="text-xs text-gray-700 dark:text-lightTeal"
    >
      3:45
    </p>
  </li>
</div>
`;
